File: Editor/Tools/BatchExecuteTool.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Threading.Tasks;
using UnityEngine;
using UnityEditor;
using McpUnity.Services;
using McpUnity.Unity;
using Newtonsoft.Json.Linq;
using Unity.EditorCoroutines.Editor;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for executing multiple operations in a single batch request.
    /// Supports sequential execution, stop-on-error, and atomic rollback outside Prefab sessions.
    /// </summary>
    public class BatchExecuteTool : McpToolBase
    {
        private readonly McpUnityServer _server;

        public BatchExecuteTool(McpUnityServer server)
        {
            _server = server;
            Name = "batch_execute";
            Description = "Executes multiple tool operations in a single batch request. Reduces " +
                          "round-trips and enables Undo-backed atomic operations outside active " +
                          "Prefab contents sessions; atomic=true is rejected while a session is active. " +
                          "Atomic rollback only restores Undo-tracked in-memory state. Asset paths " +
                          "observed during the batch window are reported, may include other editor " +
                          "activity, and have no disk-reversion guarantee from Unity Undo.";
            IsAsync = true;
        }

        public override void ExecuteAsync(JObject parameters, TaskCompletionSource<JObject> tcs)
        {
            EditorCoroutineUtility.StartCoroutineOwnerless(ExecuteBatchCoroutine(parameters, tcs));
        }

        private IEnumerator ExecuteBatchCoroutine(JObject parameters, TaskCompletionSource<JObject> tcs)
        {
            JArray operations = parameters["operations"] as JArray;
            bool stopOnError = parameters["stopOnError"]?.ToObject<bool?>() ?? true;
            bool atomic = parameters["atomic"]?.ToObject<bool?>() ?? false;

            if (atomic && !stopOnError)
            {
                tcs.SetResult(McpUnitySocketHandler.CreateErrorResponse(
                    "atomic requires stopOnError to be true.",
                    "validation_error"
                ));
                yield break;
            }

            // Validate operations array
            if (operations == null || operations.Count == 0)
            {
                tcs.SetResult(McpUnitySocketHandler.CreateErrorResponse(
                    "The 'operations' array is required and must contain at least one operation.",
                    "validation_error"
                ));
                yield break;
            }

            // Validate max operations (prevent abuse)
            if (operations.Count > 100)
            {
                tcs.SetResult(McpUnitySocketHandler.CreateErrorResponse(
                    "Maximum of 100 operations allowed per batch.",
                    "validation_error"
                ));
                yield break;
            }

            if (atomic && PrefabSessionScope.HasActiveSession)
            {
                tcs.SetResult(McpUnitySocketHandler.CreateErrorResponse(
                    "atomic=true is not supported while a Prefab contents session is active " +
                    "because preview-scene create and delete operations intentionally bypass " +
                    "Unity Undo and therefore cannot be rolled back reliably. Save or discard " +
                    "the session, or retry with atomic=false.",
                    "validation_error"
                ));
                yield break;
            }

            if (atomic)
            {
                foreach (JToken operationToken in operations)
                {
                    if (operationToken is JObject operation
                        && operation["tool"]?.ToString() == "open_prefab_contents")
                    {
                        tcs.SetResult(McpUnitySocketHandler.CreateErrorResponse(
                            "atomic=true cannot include open_prefab_contents because that operation " +
                            "would activate a Prefab contents session whose preview-scene changes " +
                            "intentionally bypass Unity Undo. Open the Prefab before a non-atomic " +
                            "batch, or run the atomic batch outside Prefab editing.",
                            "validation_error"
                        ));
                        yield break;
                    }
                }
            }

            JArray results = new JArray();
            int succeeded = 0;
            int failed = 0;
            int undoGroup = -1;
            int assetWriteCollectionId = -1;
            string[] unrevertedAssetWrites = Array.Empty<string>();

            // Start undo group for atomic operations
            if (atomic)
            {
                Undo.IncrementCurrentGroup();
                undoGroup = Undo.GetCurrentGroup();
                Undo.SetCurrentGroupName("Batch Execute");
                assetWriteCollectionId = AtomicBatchAssetWriteTracker.Begin();
            }

            try
            {
            for (int i = 0; i < operations.Count; i++)
            {
                JObject operation = operations[i] as JObject;
                if (operation == null)
                {
                    results.Add(CreateOperationResult(i, null, false, null, "Invalid operation format"));
                    failed++;

                    if (stopOnError)
                    {
                        unrevertedAssetWrites = RevertIfAtomic(
                            atomic, undoGroup, assetWriteCollectionId);
                        break;
                    }
                    continue;
                }

                string toolName = operation["tool"]?.ToString();
                JObject toolParams = operation["params"] as JObject ?? new JObject();
                string operationId = operation["id"]?.ToString() ?? i.ToString();

                // Validate tool name
                if (string.IsNullOrEmpty(toolName))
                {
                    results.Add(CreateOperationResult(i, operationId, false, null, "Missing 'tool' name in operation"));
                    failed++;

                    if (stopOnError)
                    {
                        unrevertedAssetWrites = RevertIfAtomic(
                            atomic, undoGroup, assetWriteCollectionId);
                        break;
                    }
                    continue;
                }

                // Prevent recursive batch execution
                if (toolName == Name)
                {
                    results.Add(CreateOperationResult(i, operationId, false, null, "Cannot nest batch_execute operations"));
                    failed++;

                    if (stopOnError)
                    {
                        unrevertedAssetWrites = RevertIfAtomic(
                            atomic, undoGroup, assetWriteCollectionId);
                        break;
                    }
                    continue;
                }

                // Get the tool
                if (!_server.TryGetTool(toolName, out McpToolBase tool))
                {
                    results.Add(CreateOperationResult(i, operationId, false, null, $"Unknown tool: {toolName}"));
                    failed++;

                    if (stopOnError)
                    {
                        unrevertedAssetWrites = RevertIfAtomic(
                            atomic, undoGroup, assetWriteCollectionId);
                        break;
                    }
                    continue;
                }

                // Execute the tool
                JObject toolResult = null;
                Exception toolException = null;

                if (tool.IsAsync)
                {
                    var toolTcs = new TaskCompletionSource<JObject>();

                    try
                    {
                        tool.ExecuteAsync(toolParams, toolTcs);
                    }
                    catch (Exception ex)
                    {
                        toolException = ex;
                    }

                    // Wait for async tool completion (yield must be outside try-catch)
                    if (toolException == null)
                    {
                        while (!toolTcs.Task.IsCompleted)
                        {
                            yield return null;
                        }

                        if (toolTcs.Task.IsFaulted)
                        {
                            toolException = toolTcs.Task.Exception?.InnerException ?? toolTcs.Task.Exception;
                        }
                        else
                        {
                            toolResult = toolTcs.Task.Result;
                        }
                    }
                }
                else
                {
                    try
                    {
                        toolResult = tool.Execute(toolParams);
                    }
                    catch (Exception ex)
                    {
                        toolException = ex;
                    }
                }

                // Process result
                if (toolException != null)
                {
                    results.Add(CreateOperationResult(i, operationId, false, null, toolException.Message));
                    failed++;

                    if (stopOnError)
                    {
                        unrevertedAssetWrites = RevertIfAtomic(
                            atomic, undoGroup, assetWriteCollectionId);
                        break;
                    }
                }
                else if (toolResult != null)
                {
                    if (toolResult["type"]?.ToString() == "image")
                    {
                        const string imageErrorCode = "IMAGE_RESULT_NOT_SUPPORTED_IN_BATCH";
                        string imageError =
                            $"{imageErrorCode}: Tool '{toolName}' returned image content, which " +
                            $"batch_execute cannot transport. Call '{toolName}' directly.";
                        if (toolResult["gameViewWindowCreated"]?.ToObject<bool?>() == true)
                        {
                            imageError += " Side effect: gameViewWindowCreated=true; Unity Undo " +
                                "cannot close this editor window.";
                        }
                        results.Add(CreateOperationResult(
                            i,
                            operationId,
                            false,
                            null,
                            imageError,
                            imageErrorCode));
                        failed++;

                        if (stopOnError)
                        {
                            unrevertedAssetWrites = RevertIfAtomic(
                                atomic, undoGroup, assetWriteCollectionId);
                            break;
                        }

                        yield return null;
                        continue;
                    }

                    // Check if the result indicates an error
                    bool isError = toolResult["error"] != null;
                    bool isSuccess = toolResult["success"]?.ToObject<bool?>() ?? !isError;

                    if (isSuccess && !isError)
                    {
                        results.Add(CreateOperationResult(i, operationId, true, toolResult, null));
                        succeeded++;
                    }
                    else
                    {
                        string errorMessage = toolResult["error"]?["message"]?.ToString()
                            ?? toolResult["message"]?.ToString()
                            ?? "Tool execution failed";
                        results.Add(CreateOperationResult(i, operationId, false, toolResult, errorMessage));
                        failed++;

                        if (stopOnError)
                        {
                            unrevertedAssetWrites = RevertIfAtomic(
                                atomic, undoGroup, assetWriteCollectionId);
                            break;
                        }
                    }
                }
                else
                {
                    results.Add(CreateOperationResult(i, operationId, false, null, "Tool returned null result"));
                    failed++;

                    if (stopOnError)
                    {
                        unrevertedAssetWrites = RevertIfAtomic(
                            atomic, undoGroup, assetWriteCollectionId);
                        break;
                    }
                }

                // Yield to allow Unity to process other events
                yield return null;
            }

            // Collapse undo group
            if (atomic && failed == 0)
            {
                if (undoGroup >= 0)
                    Undo.CollapseUndoOperations(undoGroup);
            }

            // Build response
            string message;
            if (failed == 0)
            {
                message = $"Successfully executed {succeeded}/{operations.Count} operations.";
            }
            else if (atomic && stopOnError)
            {
                message = "Batch execution failed. Unity Undo restored only Undo-tracked " +
                    $"in-memory state. {succeeded} operations succeeded before failure.";
                if (unrevertedAssetWrites.Length > 0)
                {
                    message += $" {unrevertedAssetWrites.Length} asset path(s) were observed " +
                        "in Unity save/postprocess callbacks during this batch's collection " +
                        "window. This evidence may include writes from other editor activity; " +
                        "Unity Undo does not establish that those disk writes were reverted. " +
                        "See unrevertedAssetWrites.";
                }
                else
                {
                    message += " No asset save/postprocess callbacks were observed during this " +
                        "batch's collection window.";
                }
            }
            else if (stopOnError)
            {
                message = $"Batch execution stopped on error. {succeeded}/{operations.Count} operations succeeded.";
            }
            else
            {
                message = $"Batch execution completed with errors. {succeeded}/{operations.Count} operations succeeded, {failed} failed.";
            }

            var response = new JObject
            {
                ["success"] = failed == 0,
                ["type"] = "text",
                ["message"] = message,
                ["results"] = results,
                ["summary"] = new JObject
                {
                    ["total"] = operations.Count,
                    ["succeeded"] = succeeded,
                    ["failed"] = failed,
                    ["executed"] = succeeded + failed
                }
            };

            if (atomic && failed > 0)
                response["unrevertedAssetWrites"] = new JArray(unrevertedAssetWrites);

            tcs.SetResult(response);
            }
            finally
            {
                if (assetWriteCollectionId >= 0)
                    AtomicBatchAssetWriteTracker.End(assetWriteCollectionId);
            }
        }

        private string[] RevertIfAtomic(
            bool atomic,
            int undoGroup,
            int assetWriteCollectionId)
        {
            if (!atomic)
                return Array.Empty<string>();

            string[] assetWrites = AtomicBatchAssetWriteTracker.End(assetWriteCollectionId);
            if (undoGroup >= 0)
                Undo.RevertAllDownToGroup(undoGroup);

            return assetWrites;
        }

        private JObject CreateOperationResult(
            int index,
            string id,
            bool success,
            JObject result,
            string error,
            string errorCode = null)
        {
            var operationResult = new JObject
            {
                ["index"] = index,
                ["id"] = id ?? index.ToString(),
                ["success"] = success
            };

            if (result != null)
            {
                operationResult["result"] = result;
            }

            if (!success)
            {
                operationResult["error"] = error ?? "Unknown error";
                if (!string.IsNullOrEmpty(errorCode))
                    operationResult["errorCode"] = errorCode;
            }

            return operationResult;
        }
    }
}

File: Server~/src/tools/batchExecuteTool.ts
import * as z from 'zod';
import { McpUnity } from '../unity/mcpUnity.js';
import { Logger } from '../utils/logger.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';
import { payloadContent } from '../utils/toolPayload.js';
import { getRegisteredToolSchema } from '../utils/structuredContentSeam.js';

const toolName = 'batch_execute';
const toolDescription = `Executes multiple tool operations in a single batch request.
Reduces network round-trips and enables Undo-backed atomic operations outside active Prefab contents sessions.
atomic=true is rejected while a Prefab contents session is active and cannot include open_prefab_contents because preview changes bypass Unity Undo.
Atomic rollback only restores Undo-tracked in-memory state. Asset paths observed during the batch window are reported, may include other editor activity, and have no disk-reversion guarantee from Unity Undo.
Inner schema validation occurs before any operation executes; with stopOnError=true, any operation validation failure prevents the entire batch from executing.
Performance improvement: 10-100x for repetitive operations.`;

const operationSchema = z.object({
  tool: z.string().describe('The name of the tool to execute'),
  params: z.record(z.string(), z.any()).optional().default({}).describe('Parameters to pass to the tool'),
  id: z.string().optional().describe('Optional identifier for this operation (for tracking in results)')
});

const paramsSchema = z.object({
  operations: z.array(operationSchema)
    .min(1, 'At least one operation is required')
    .max(100, 'Maximum of 100 operations allowed per batch')
    .describe('Array of operations to execute sequentially'),
  stopOnError: z.boolean()
    .default(true)
    .describe('If true, stops execution on the first error. Default: true'),
  atomic: z.boolean()
    .default(false)
    .describe('If true, restores Undo-tracked in-memory state if any operation fails. Asset paths observed during the batch window are reported, may include other editor activity, and have no disk-reversion guarantee from Unity Undo. Rejected while a Prefab contents session is active and cannot include open_prefab_contents. Default: false')
});

/**
 * Result of a single operation in the batch
 */
interface OperationResult {
  index: number;
  id: string;
  success: boolean;
  result?: any;
  error?: string;
  errorCode?: string;
}

/**
 * Summary of batch execution
 */
interface BatchSummary {
  total: number;
  succeeded: number;
  failed: number;
  executed: number;
}

/**
 * Response from the batch execute tool
 */
interface BatchExecuteResponse {
  success: boolean;
  type: string;
  message: string;
  results: OperationResult[];
  summary: BatchSummary;
  unrevertedAssetWrites?: string[];
}

const validationErrorMessage = (
  index: number,
  id: string,
  tool: string,
  reason: string,
): string => `Batch operation ${index} (id "${id}", tool "${tool}") ${reason}`;

const pickParsedCallerParams = (
  parsedParams: Record<string, unknown>,
  callerParams: Record<string, unknown>,
): Record<string, unknown> => Object.fromEntries(
  Object.keys(callerParams)
    .filter((key) => Object.prototype.hasOwnProperty.call(parsedParams, key))
    .map((key) => [key, parsedParams[key]]),
);

const throwWithNodeSideContext = (
  error: unknown,
  locallyRejectedOperations: OperationResult[],
  validationWarnings: string[],
): never => {
  if (locallyRejectedOperations.length === 0 && validationWarnings.length === 0) {
    throw error;
  }

  const type = error instanceof McpUnityError ? error.type : ErrorType.INTERNAL;
  const originalMessage = error instanceof Error ? error.message : String(error);
  const contextLines = [
    ...(locallyRejectedOperations.length === 0
      ? []
      : [
          'Node-side rejections:',
          ...locallyRejectedOperations.map((operation) => `  - ${operation.error}`),
        ]),
    ...(validationWarnings.length === 0
      ? []
      : [
          'Warnings:',
          ...validationWarnings.map((warning) => `  - ${warning}`),
        ]),
  ];
  const originalDetails = error instanceof McpUnityError ? error.details : undefined;
  const details = originalDetails !== null
    && typeof originalDetails === 'object'
    && !Array.isArray(originalDetails)
    ? originalDetails
    : originalDetails === undefined
      ? {}
      : { unityErrorDetails: originalDetails };

  throw new McpUnityError(
    type,
    `${originalMessage}\n\n${contextLines.join('\n')}`,
    {
      ...details,
      locallyRejectedOperations,
      validationWarnings,
    },
  );
};

/**
 * Creates and registers the Batch Execute tool with the MCP server
 */
export function registerBatchExecuteTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);

  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, {
          operationCount: params.operations?.length,
          stopOnError: params.stopOnError,
          atomic: params.atomic
        });
        const result = await batchExecuteHandler(server, mcpUnity, params, logger);
        logger.info(`Tool execution completed: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

async function batchExecuteHandler(
  server: McpServer,
  mcpUnity: McpUnity,
  params: z.infer<typeof paramsSchema>,
  logger: Logger
): Promise<CallToolResult> {
  // Validate operations array
  if (!params.operations || params.operations.length === 0) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "The 'operations' array is required and must contain at least one operation"
    );
  }

  if (params.operations.length > 100) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "Maximum of 100 operations allowed per batch"
    );
  }

  const stopOnError = params.stopOnError ?? true;
  const forwardedOperationIndexes: number[] = [];
  const locallyRejectedOperations: OperationResult[] = [];
  const validationWarnings: string[] = [];
  const forwardedOperations = params.operations.flatMap((op, index) => {
    const id = op.id ?? index.toString();
    let validationFailure: string | undefined;

    if (op.tool === toolName) {
      validationFailure = validationErrorMessage(
        index,
        id,
        op.tool,
        'Cannot nest batch_execute operations',
      );
    }

    const schema = getRegisteredToolSchema(server, op.tool);
    if (!validationFailure && !schema) {
      const warning = validationErrorMessage(
        index,
        id,
        op.tool,
        'was not validated by Node because its schema is not registered; Unity will perform authoritative tool lookup and validation.',
      );
      validationWarnings.push(warning);
      logger.warn(warning);
    }

    const validation = validationFailure || !schema
      ? undefined
      : schema.safeParse(op.params ?? {});
    if (validation && !validation.success) {
      const details = validation.error.issues
        .map((issue) => issue.path.length > 0
          ? `${issue.message} at ${issue.path.join('.')}`
          : issue.message)
        .join('; ');
      validationFailure = validationErrorMessage(
        index,
        id,
        op.tool,
        `has invalid params: ${details}`,
      );
    }

    if (validationFailure) {
      if (stopOnError) {
        throw new McpUnityError(ErrorType.VALIDATION, validationFailure);
      }

      locallyRejectedOperations.push({
        index,
        id,
        success: false,
        error: validationFailure,
        errorCode: 'NODE_SCHEMA_VALIDATION',
      });
      return [];
    }

    const parsedParams = validation?.success
      ? validation.data as Record<string, unknown>
      : undefined;
    const callerParams = op.params ?? {};
    forwardedOperationIndexes.push(index);
    return [{
      tool: op.tool,
      // Keep caller presence semantics while matching direct-call coercion.
      // A caller-provided top-level object retains any parsed nested defaults.
      params: parsedParams
        ? pickParsedCallerParams(parsedParams, callerParams)
        : callerParams,
      id,
    }];
  });

  logger.info(`Sending batch with ${forwardedOperations.length} operations to Unity`);

  let response: BatchExecuteResponse;
  if (forwardedOperations.length === 0) {
    response = {
      success: false,
      type: 'text',
      message: `Batch execution completed with errors. 0/${params.operations.length} operations succeeded, ${locallyRejectedOperations.length} failed.`,
      results: locallyRejectedOperations,
      summary: {
        total: params.operations.length,
        succeeded: 0,
        failed: locallyRejectedOperations.length,
        executed: 0,
      },
    };
  } else {
    const unityResponse = await mcpUnity.sendRequest({
      method: toolName,
      params: {
        operations: forwardedOperations,
        stopOnError,
        atomic: params.atomic ?? false
      }
    })
      .catch((error) => throwWithNodeSideContext(
        error,
        locallyRejectedOperations,
        validationWarnings,
      )) as BatchExecuteResponse;

    const rawUnityResults = Array.isArray(unityResponse.results)
      ? unityResponse.results
      : [];
    if (!stopOnError) {
      const receivedIndexes = rawUnityResults.map((result) => result.index);
      const receivedIndexSet = new Set(receivedIndexes);
      const missingForwardedIndexes = forwardedOperationIndexes.filter(
        (_originalIndex, forwardedIndex) => !receivedIndexSet.has(forwardedIndex),
      );
      const hasUnexpectedIndexes = receivedIndexes.some((index) =>
        !Number.isInteger(index) || index < 0 || index >= forwardedOperations.length);
      const hasDuplicateIndexes = receivedIndexSet.size !== receivedIndexes.length;

      if (
        rawUnityResults.length !== forwardedOperations.length
        || missingForwardedIndexes.length > 0
        || hasUnexpectedIndexes
        || hasDuplicateIndexes
      ) {
        const protocolError = new McpUnityError(
          ErrorType.TOOL_EXECUTION,
          'Unity batch response violated result coverage for stopOnError=false: ' +
          `expected ${forwardedOperations.length} results but received ${rawUnityResults.length}; ` +
          `missing original operation indexes: ${missingForwardedIndexes.join(', ') || 'none'}.`,
          {
            expectedResultCount: forwardedOperations.length,
            actualResultCount: rawUnityResults.length,
            missingOriginalOperationIndexes: missingForwardedIndexes,
            receivedUnityResultIndexes: receivedIndexes,
          },
        );
        throwWithNodeSideContext(
          protocolError,
          locallyRejectedOperations,
          validationWarnings,
        );
      }
    }

    if (locallyRejectedOperations.length === 0) {
      response = unityResponse;
    } else {
      const unityResults = rawUnityResults.map((result, resultIndex) => ({
        ...result,
        index: forwardedOperationIndexes[result.index]
          ?? forwardedOperationIndexes[resultIndex]
          ?? result.index,
      }));
      const results = [...unityResults, ...locallyRejectedOperations]
        .sort((left, right) => left.index - right.index);
      const succeeded = unityResponse.summary?.succeeded
        ?? unityResults.filter((result) => result.success).length;
      const unityFailed = unityResponse.summary?.failed
        ?? unityResults.filter((result) => !result.success).length;
      const failed = unityFailed + locallyRejectedOperations.length;
      const executed = unityResponse.summary?.executed ?? unityResults.length;

      response = {
        ...unityResponse,
        success: false,
        message: `Batch execution completed with errors. ${succeeded}/${params.operations.length} operations succeeded, ${failed} failed.`,
        results,
        summary: {
          total: params.operations.length,
          succeeded,
          failed,
          executed,
        },
      };
    }
  }

  // Format the response message
  let resultText = response.message || 'Batch execution completed';

  // Add summary details
  if (response.summary) {
    resultText += `\n\nSummary: ${response.summary.succeeded}/${response.summary.total} succeeded`;
    if (response.summary.failed > 0) {
      resultText += `, ${response.summary.failed} failed`;
    }
  }

  if (validationWarnings.length > 0) {
    resultText += '\n\nWarnings:\n' + validationWarnings
      .map((warning) => `  - ${warning}`)
      .join('\n');
  }

  // Build structured results with full tool data for each operation
  const structuredResults = response.results?.map((res: OperationResult) => {
    if (res.result?.type === 'image') {
      const genericError =
        'Image results are not supported in batch_execute; call the image-producing tool directly.';
      const unitySideEffect = typeof res.error === 'string'
        && res.error.includes('gameViewWindowCreated=true')
        ? res.error
        : res.result.gameViewWindowCreated === true
          ? 'Side effect: gameViewWindowCreated=true; Unity Undo cannot close this editor window.'
          : undefined;
      return {
        id: res.id,
        status: 'Error',
        errorCode: 'IMAGE_RESULT_NOT_SUPPORTED_IN_BATCH',
        error: unitySideEffect ? `${genericError} ${unitySideEffect}` : genericError
      };
    }

    const entry: Record<string, any> = {
      id: res.id,
      status: res.success ? 'OK' : 'Error'
    };
    if (res.result !== undefined) {
      entry.data = res.result;
    }
    if (!res.success && res.error) {
      entry.error = res.error;
    }
    if (!res.success && res.errorCode) {
      entry.errorCode = res.errorCode;
    }
    return entry;
  }) ?? [];

  if (structuredResults.some((result) =>
    result.errorCode === 'IMAGE_RESULT_NOT_SUPPORTED_IN_BATCH')) {
    resultText += '\n\nIMAGE_RESULT_NOT_SUPPORTED_IN_BATCH: Call the image-producing tool directly.';
  }

  // Include full results JSON so AI clients can access each tool's return data
  const structuredPayload = {
    message: response.message
      || `${response.summary?.succeeded ?? 0}/${response.summary?.total ?? structuredResults.length} operations succeeded`,
    results: structuredResults,
    summary: response.summary,
    ...(response.unrevertedAssetWrites === undefined
      ? {}
      : { unrevertedAssetWrites: response.unrevertedAssetWrites }),
    ...(validationWarnings.length === 0 ? {} : { warnings: validationWarnings }),
  };

  const hasOperationFailures = !response.success
    || structuredResults.some((result) => result.status === 'Error');

  const result: CallToolResult = {
    content: [
      {
        type: 'text',
        text: resultText
      },
      payloadContent(structuredPayload)
    ]
  };

  if (hasOperationFailures) {
    result.isError = true;
  }

  return result;
}
